diesel/src/query_builder/upsert/conditionally_insert_statement.rs

```diff
@@ -0,0 +1,108 @@
+use std::fmt::Debug;
+
+use crate::backend::{Backend, DieselReserveSpecialization};
+use crate::query_builder::upsert::on_conflict_clause::OnConflictValues;
+use crate::query_builder::where_clause::{NoWhereClause, WhereAnd};
+use crate::query_builder::{
+    where_clause, AsQuery, AstPass, InsertStatement, QueryFragment, QueryId,
+};
+use crate::query_dsl::filter_dsl::FilterDsl;
+use crate::sql_types::BoolOrNullableBool;
+use crate::{Expression, QueryResult, QuerySource, RunQueryDsl};
+
+#[derive(Debug, Clone)]
+pub struct ConditionallyInsertStatement<Stmt, WhereClause = NoWhereClause> {
+    insert: Stmt,
+    where_clause: WhereClause,
+}
+
+impl<Stmt, WhereClause> ConditionallyInsertStatement<Stmt, WhereClause> {
+    pub fn new(
+        insert: Stmt,
+        where_clause: WhereClause,
+    ) -> ConditionallyInsertStatement<Stmt, WhereClause> {
+        ConditionallyInsertStatement {
+            insert,
+            where_clause,
+        }
+    }
+}
+
+impl<Stmt, WhereClause> QueryId for ConditionallyInsertStatement<Stmt, WhereClause>
+where
+    Stmt: QueryId,
+{
+    type QueryId = Stmt::QueryId;
+    const HAS_STATIC_QUERY_ID: bool = Stmt::HAS_STATIC_QUERY_ID;
+}
+
+impl<Stmt, WhereClause, DB> QueryFragment<DB> for ConditionallyInsertStatement<Stmt, WhereClause>
```

Comment:
As far as I understand that's a postgres only feature, right? In that case we should provide only an `QueryFragment` implementation for `Pg` instead of a generic one for all database backend.